include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -1752,656 +1572,197 @@ namespace internal
+    const Number *input1 = input + n_blocks[0] * n_blocks[1];
+    const Number *input2 = input1 + n_blocks[0] * n_blocks[1];
+    for (int i2 = 0; i2 < n_blocks[1]; ++i2)
       {
-        for (int i1 = 0; i1 < n_blocks1; ++i1)
-          {
-            if (contract_onto_face == true)
-              {
-                Number res0 = shape_values[0] * in[0];
-                Number res1, res2;
-
+        for (int i1 = 0; i1 < n_blocks[0]; ++i1)
+          {
+            const Number in = input[i1];
+            Number       in1, in2;
+            if (max_derivative > 0)
+              in1 = input1[i1];
+            if (max_derivative > 1)
+              in2 = input2[i1];
+            for (int col = 0; col < n_rows; ++col)
+              {
+                Number result =
+                  add ? (output[col * stride] + shape_values[col] * in) :
+                        (shape_values[col] * in);
                 if (max_derivative > 0)
-                  res1 = shape_values[n_rows] * in[0];
-
+                  result += shape_values[col + n_rows] * in1;
                 if (max_derivative > 1)
-                  res2 = shape_values[2 * n_rows] * in[0];
+                  result += shape_values[col + 2 * n_rows] * in2;
 
-                for (int ind = 1; ind < n_rows; ++ind)
-                  {
-                    res0 += shape_values[ind] * in[in_stride * ind];
-                    if (max_derivative > 0)
-                      res1 += shape_values[ind + n_rows] * in[in_stride * ind];
-
-                    if (max_derivative > 1)
-                      res2 +=
-                        shape_values[ind + 2 * n_rows] * in[in_stride * ind];
-                  }
-                if (add)
-                  {
-                    out[0] += res0;
-
-                    if (max_derivative > 0)
-                      out[out_stride] += res1;
-
-                    if (max_derivative > 1)
-                      out[2 * out_stride] += res2;
-                  }
-                else
-                  {
-                    out[0] = res0;
-
-                    if (max_derivative > 0)
-                      out[out_stride] = res1;
-
-                    if (max_derivative > 1)
-                      out[2 * out_stride] = res2;
-                  }
-              }
-            else
-              {
-                for (int col = 0; col < n_rows; ++col)
-                  {
-                    if (add)
-                      out[col * in_stride] += shape_values[col] * in[0];
-                    else
-                      out[col * in_stride] = shape_values[col] * in[0];
-
-                    if (max_derivative > 0)
-                      out[col * in_stride] +=
-                        shape_values[col + n_rows] * in[out_stride];
-
-                    if (max_derivative > 1)
-                      out[col * in_stride] +=
-                        shape_values[col + 2 * n_rows] * in[2 * out_stride];
-                  }
-              }
```

Comment:
What happened to all the if-else statements? Are these not needed anymore?

Reply:
These are now gone because I use two separate functions via `std::enable_if` for `interpolate_to_face` and the transpose (which interpolates from the face). The cases in incrementing the pointers are much simplified via the `n_blocks` and `steps` variables, that get controlled in the evaluation kernels (with different values of these variables I can both implement the code for Raviart-Thomas and isotropic tensor products).